### PR TITLE
Remove cyclic reference between space and page resource (merge after PR #268)

### PR DESCRIPTION
**Pull request by qinsoon**
This PR focuses removing the cyclic references between `Space` and `PageResource`. By refactoring our code base to achieve such a goal, a lot of unsafe code is also removed. This PR works towards 1) https://github.com/mmtk/mmtk-core/issues/59 by tidying up sync/unsync, 2) https://github.com/mmtk/mmtk-core/issues/57 by removing the cyclic references, and 3) https://github.com/mmtk/mmtk-core/issues/97 by removing a lot of unsafe code of turning immutable refs into mutable refs and unsafely implementing `Sync` for certain types.

Changes:
* Removed the cyclic reference between `Space` and `PageResource` by
  * Moved the `head_discontiguous_region` field and related methods from `CommonSpace` to `CommonPageResource` so `PageResource` can update the field without needing a reference to `Space`. `head_discontiguous_region` is now protected by a `Mutex` (The mutex can be removed but instead we would need unsafe code and I think it is hard to argue the unsafe code is correct).
  * Allocating with `PageResource` now returns a result, and `Space` will do some space related work based on the result (such as growing space, ensuring mmapped and zeroing) which used to be done in `PageResource`.
* Removed `unsafe_common_mut()` in `Space` which is no longer needed, which was needed to access `head_discontiguous_region`.
* Removed `UnsafeCell` for `CommonSpace` in a few space implementations. Removed `unsafe impl Sync` for those spaces.
* Removed a few other unsafe code that is no longer needed for those changes.

Performance:

In the refactoring, this PR fixes a performance bug we have in mmtk-core that we hold a page resource lock while zeroing memory, which is different from Java MMTk and results in a slowdown. Reported by our CI, this PR improves performance by 2%-5% for some Dacapo benchmarks for both JikesRVM and OpenJDK, and it improves micro benchmarks such as binary trees and multi threaded GC bench for OpenJDK by 30%-40%. 

The only benchmark that showed a slowdown in the CI is `hsqldb` for GenCopy on OpenJDK (roughly 2-4% slowdown). I think it is a false positive. We have roughly 10 out of 40 invocations failed due to failing to converge the result. We run Dacapo benchmarks with `converge`, which runs the benchmarks as warmup until the result converges to a fixed point. Before this PR (the current master), we observe the execution times of iterations for hsqldb are sometimes like these, 370ms, 340ms, 370ms, 340ms, and this leads to a converge fail. But after this PR, we haven't seen any converge failure. The possible reason is that the page resource lock is more contended in some runs and less contended in others. More contended lock leads to a worse case, and less contended leads to a better case. Among the runs, it is more likely for the results to converge to better cases, and fail if the results are mixed. Thus before this PR, the results from the master tend to be better in our CI measurement. 

I also ran `hsqldb` with 5 iterations (no converge flag) 80 invocations, comparing the master and the PR. The result is as below and it shows a speedup (clearly not a slowdown). 
WeChat Screenshot_20210416122622

### CODE REVIEW

**Review by javadamiri — APPROVED** [review]
LGTM

**Review by wenyuzhao — APPROVED** [review]
LGTM. Thanks for the PR. This simplifies the code quite a lot.
---
BTW, recently I noticed a rust nightly API to create cyclic references without unsafe: `Arc::new_cyclic` https://doc.rust-lang.org/std/sync/struct.Arc.html#method.new_cyclic

File: src/policy/space.rs [hunk 1]
```diff
@@ -266,21 +268,42 @@ pub trait Space<VM: VMBinding>: 'static + SFT + Sync + Downcast {
...
+                    // are no longer guarded by a lock. If we see any issue here, considering
+                    // adding a space lock here.
+                    let bytes = conversions::pages_to_bytes(res.pages);
+                    self.grow_space(res.start, bytes, res.new_chunk);
+                    self.common().mmapper.ensure_mapped(
+                        res.start,
+                        res.pages,
+                        VM::VMActivePlan::global().global_side_metadata_specs(),
+                        self.local_side_metadata_specs(),
+                    );
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply src/policy/space.rs:285]: Yes.

**Comment by qinsoon** [comment]
> ## LGTM. Thanks for the PR. This simplifies the code quite a lot.
> BTW, recently I noticed a rust nightly API to create cyclic references without unsafe: `Arc::new_cyclic` https://doc.rust-lang.org/std/sync/struct.Arc.html#method.new_cyclic

Thanks for the info. Good to know that Rust has some support for cyclic references.

**Pull request merged**